text/0000-const-static-type-elision.md

```diff
@@ -0,0 +1,153 @@
+inference, rather than attempting to infer a type based on a later use of the
+item as with `let`-bound variables. For example, the following would result in a
+type error, because there are multiple possible types for the literal 42
+(e.g. `i16`, `i32`, etc.), even though the use in `get_big_number` would require
+it to be `i64`.
+
+```rust
+const THE_ANSWER = 42; // nothing in RHS indicates this must be i64
+
+fn get_big_number() -> i64 {
+    THE_ANSWER
+}
+```
+
+## Integer/Float Fallback
+
+The fallback rules (specifically, defaulting integer literals to `i32` and float
+literals to `f64`) are disallowed in cases where multiple typings are valid to
+prevent the type of an exported item from changing only by removing a type
+annotation. For example, say some crate exports the following:
+
+```rust
+const X: i64 = 5;
+```
+
+If the developer later decides to elide the type annotation, then fallback would
+infer the type of `X` as `i32` rather than `i64`. If `X` is exported but not
+used within the crate, then this change could break downstream code without the
+crate developer realizing it. Admittedly, that scenario is unlikely, but
+ruling out fallback is the most conservative option and could always be added
+back in later.
+
+Fallback is acceptable, however, if the overall type is still unique even
+without the fallback rules, as in this example:
+
+```rust
+const fn foo<T>(_: T) -> char { 'a' }
+const X = foo(22);
+```
+
+## Closures
+
+This design would allow closures (rather than just references to closures) to be
+used as `const`/`static` items, because the programmer no longer has to write
+down an inexpressible type. This shouldn't pose any particular difficulties from
+an implementation perspective, but it's worth being aware of.
+
+Documentation projects such as rustdoc may need to deal with this as a special
+case. @withoutboats
+[suggests](https://internals.rust-lang.org/t/pre-rfc-elide-type-annotations-from-const-and-static-items/5175/2?u=jschuster)
+coercing closures to fn types as one possible solution.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+_The Rust Reference_ should record the rules for when the annotation is
+optional. _The Rust Programming Language_ and _Rust by Example_ should remove
+the sections that say annotations are required, and they may want to consider
+removing annotations from their examples of `const` and `static` items (see
+"Unresolved questions" below).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Some users may find it more difficult to understand large constant expressions
+  without a type annotation. Better IDE support for inferred types would help
+  mitigate this issue.
+* Const functions may make it more difficult for a programmer to infer the type
+  of a const/static item just be reading it. Most likely, though, most uses of
+  const functions in this context will be things like `AtomicUsize::new(0)`
+  where the type is obvious.
```

Comment:
True, they can be used in lots of ways, but since the motivation for [that RFC](https://github.com/rust-lang/rfcs/pull/911) was to initialize types like `AtomicUsize<T>` and `Cell<T>` in constant contexts, I'm assuming that will be the typical use-case. I could be wrong, though; perhaps people who were involved in that RFC discussion could provide more info.

In those more complicated cases where documenting the type is important for readability, the programmer should leave the type in. But mandating that practice by requiring annotations on all `const`/`static` items, even the simple ones, seems like overkill to me. I'll add a note about this to the RFC.